src/network/networkd-sysctl.c

```diff
@@ -4,18 +4,209 @@
 #include <linux/if.h>
 #include <linux/if_arp.h>
 
+#include "sd-messages.h"
+
 #include "af-list.h"
+#include "cgroup-util.h"
+#include "fd-util.h"
 #include "missing_network.h"
 #include "networkd-link.h"
 #include "networkd-lldp-tx.h"
 #include "networkd-manager.h"
 #include "networkd-ndisc.h"
 #include "networkd-network.h"
 #include "networkd-sysctl.h"
+#include "path-util.h"
 #include "socket-util.h"
 #include "string-table.h"
 #include "sysctl-util.h"
 
+#if HAVE_VMLINUX_H
+
+#include "bpf-link.h"
+
+#include "bpf/sysctl_monitor/sysctl-monitor-skel.h"
+#include "bpf/sysctl_monitor/sysctl-write-event.h"
+
+static struct sysctl_monitor_bpf *sysctl_monitor_bpf_free(struct sysctl_monitor_bpf *obj) {
+        sysctl_monitor_bpf__destroy(obj);
+        return NULL;
+}
+
+static struct ring_buffer *rb_free(struct ring_buffer *rb) {
+        sym_ring_buffer__free(rb);
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct sysctl_monitor_bpf *, sysctl_monitor_bpf_free);
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct ring_buffer *, rb_free);
+
+static int sysctl_event_handler(void *ctx, void *data, size_t data_sz) {
+        struct sysctl_write_event *we = ASSERT_PTR(data);
+        Hashmap **sysctl_shadow = ASSERT_PTR(ctx);
+        _cleanup_free_ char *path = NULL;
+        char *value;
+
+        /* Returning a negative value interrupts the ring buffer polling,
+         * so do it only in case of a fatal error like a version mismatch. */
+        if (we->version != 1)
+                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                "Unexpected sysctl event, disabling sysctl monitoring: %d", we->version);
+
+        if (we->errorcode < 0) {
+                log_warning_errno(we->errorcode, "Sysctl monitor BPF returned error: %m");
+                return 0;
+        }
+
+        path = path_join("/proc/sys", we->name);
+        if (!path) {
+                log_oom();
+                return 0;
+        }
+
+        /* If we never managed this handle, ignore it. */
+        value = hashmap_get(*sysctl_shadow, path);
+        if (!value)
+                return 0;
+
+        if (!strneq(value, we->newvalue, sizeof(we->newvalue)))
+                log_struct(LOG_WARNING,
+                           "MESSAGE_ID=" SD_MESSAGE_SYSCTL_CHANGED_STR,
+                           "COMM_PID=%d", we->pid,
```

Comment:
please use PID_FMT

Reply:
This is weird, if I just include `format-util.h`, I get this build error:
```
In file included from ../src/network/networkd-sysctl.c:4:
../src/basic/linux/if.h:110:41: error: redeclaration of enumerator ‘IFF_UP’
  110 | #define IFF_UP                          IFF_UP
      |                                         ^~~~~~
../src/basic/linux/if.h:83:9: note: previous definition of ‘IFF_UP’ with type ‘enum net_device_flags’
   83 |         IFF_UP                          = 1<<0,  /* sysfs */
      |         ^~~~~~
../src/basic/linux/if.h:111:41: error: redeclaration of enumerator ‘IFF_BROADCAST’
  111 | #define IFF_BROADCAST                   IFF_BROADCAST
      |                                         ^~~~~~~~~~~~~
../src/basic/linux/if.h:84:9: note: previous definition of ‘IFF_BROADCAST’ with type ‘enum net_device_flags’
   84 |         IFF_BROADCAST                   = 1<<1,  /* __volatile__ */
      |         ^~~~~~~~~~~~~
...
In file included from ../src/basic/format-util.h:5,
                 from ../src/network/networkd-sysctl.c:12:
/usr/include/net/if.h:111:8: error: redefinition of ‘struct ifmap’
  111 | struct ifmap
      |        ^~~~~
../src/basic/linux/if.h:194:8: note: originally defined here
  194 | struct ifmap {
      |        ^~~~~
/usr/include/net/if.h:126:8: error: redefinition of ‘struct ifreq’
  126 | struct ifreq
      |        ^~~~~
../src/basic/linux/if.h:232:8: note: originally defined here
  232 | struct ifreq {
      |        ^~~~~
/usr/include/net/if.h:176:8: error: redefinition of ‘struct ifconf’
  176 | struct ifconf
      |        ^~~~~~
../src/basic/linux/if.h:284:8: note: originally defined here
  284 | struct ifconf  {
      |        ^~~~~~
```

Reply:
yeah kernel and glibc network headers famously conflict.it's a mess.